### PR TITLE
Throw error if unicode is in reserved space

### LINKED ISSUE #338: Escaped unicode characters are encoded incorrectly
In strings with espaced Unicode characters (i.e. `"\uABCD"`), characters in the range from `D800` to `DFFF` always get encoded as `FFFD` (65533 decimal).

Here's from the F# Interactive:

```
    > "\uD500".[0] |> uint16 ;;
    val it : uint16 = 54528us
    > "\uD700".[0] |> uint16 ;;
    val it : uint16 = 55040us
    > "\uD800".[0] |> uint16 ;;
    val it : uint16 = 65533us
    > "\uD900".[0] |> uint16 ;;
    val it : uint16 = 65533us
    > "\uE000".[0] |> uint16 ;;
    val it : uint16 = 57344us
```

Here's an SO question that prompted me to fiddle with it: http://stackoverflow.com/questions/29359408/surrogate-pair-detection-fails

Here's another issue I've (mistakenly?) opened in the *other* repo: https://github.com/fsharp/fsharp/issues/399

### PR DESCRIPTION
**Pull request by forki**
This fixes #338

image

### CODE REVIEW

File: src/fsharp/lex.fsl [hunk 1]
```diff
@@ -752,8 +752,12 @@ and string sargs skip = parse
       
  | unicodeGraphShort
     { let (buf,_fin,m,args) = sargs 
-      addUnicodeChar buf (int (unicodeGraphShort (lexemeTrimLeft lexbuf 2)));
-      if not skip then (STRING_TEXT (LexCont.String(!args.ifdefStack,m)))  else string sargs skip lexbuf  }
+      let c = int (unicodeGraphShort (lexemeTrimLeft lexbuf 2))
+      if c >= 0xD800 && c <= 0xD8FF then
```
Comment (forki) [inline src/fsharp/lex.fsl]: I have to change 0xD8FF to 0xDFFF

**Comment by forki** [comment]
dear @msftclas bot, I already signed the CLA in Codeplex times.

**Comment by forki** [comment]
Conformance\LexicalAnalysis\StringsAndCharacters (ByteString03.fs) -- failed
    Conformance\LexicalAnalysis\StringsAndCharacters (TripleQuoteString01.fs) -- failed
    Conformance\LexicalAnalysis\StringsAndCharacters (TripleQuoteStringInFSI02.fsx) -- failed

3 unrelated tests failed on appveyor.

**Comment by latkin** [comment]
Test surely failed due to ongoing line-ending issues with AppVeyor, no worries there.

**Comment by forki** [comment]
Closed in favor of #348

**Pull request closed without merging**